README.md
# ![LOCKJAW](docs/ljlogo192.png) About

*   [Home](http://www.pineight.com/lj/)
*   About
*   [Download](http://www.pineight.com/lj/dl)

<div id="content">

[Installing](#installing) | [Skinning](#skin) | [Controls](#controls) | [Scoring](#scoring) | [Gimmicks](#gimmicks) | [Options](#options) | [Discussion](#discussion) | [Legal](#legal)

LOCKJAW is a [free software](http://www.gnu.org/philosophy/free-sw.html) implementation of the the Soviet Mind Game, a highly popular computer puzzle game that involves guiding tetrominoes into neat stacks in a well. This game was designed in the mid-1980s by Russian game designer Alexey Pajitnov and was first implemented in a software product called [TETRIS](http://en.wikipedia.org/wiki/Tetris)®.[*](#trademarks) Other products implementing the Soviet Mind Game include [Quadra](http://quadra.sourceforge.net/), [Abandoned Bricks](http://abrick.sourceforge.net/), [KSirtet](http://ksirtet.sourceforge.net/), [F.B.G.](http://fbg.sourceforge.net/), [N-Blox](http://www.neave.com/games/nblox/), and [Heboris](http://www.tetrisconcept.com/forum/viewtopic.php?t=54). While originally developed to parody the behavior of a few notorious implementations of the Soviet Mind Game, LOCKJAW is now a platform for research into the properties of the game, into the effects of rule variations, and into the capabilities of the human mind to react.

[Tetrominoes](http://en.wikipedia.org/wiki/Tetromino) are geometric shapes made of four connected square blocks. There are seven distinct tetrominoes, shaped roughly like letters of the Latin alphabet:

![(Illustration of the seven tetrominoes)](docs/ijlo-stz.png)  
Tetrominoes. Top row: I, J, L, O. Bottom row: S, T, Z.

Some products spell the term as "tetramino", "tetrimino", or "tetrad".

<div class="screenshot">![(Game screenshot)](docs/ljsnap018.png)

LOCKJAW gameplay screen. Hover over objects to see their names.

<map name="ljsnap018_map"><area shape="rect" coords="12,12,48,36" alt="Hold piece" title="Hold piece" style="border: 1px solid white"> <area shape="rect" coords="48,0,192,36" alt="Status" title="Status"> <area shape="rect" coords="96,48,132,108" alt="Falling piece" title="Falling piece"> <area shape="rect" coords="96,168,132,216" alt="Shadow" title="Shadow"> <area shape="rect" coords="12,36,132,276" alt="Well" title="Well"> <area shape="rect" coords="36,276,108,300" alt="Game mode" title="Game mode"> <area shape="rect" coords="144,36,192,252" alt="Next pieces" title="Next pieces"> <area shape="rect" coords="216,60,360,240" alt="Gus" title="Gus"></map> </div>

The well is 10 blocks wide by 20 rows high. (There are four out-of-bounds rows above the top of the visible portion for a total of 24 rows.) The player can rotate and shift the tetrominoes as they fall in order to pack them tightly into the well. If a tetromino lands on the floor or other blocks and remains motionless for half a second, it locks into place, and the next tetromino begins to fall. The next few tetrominoes to fall are displayed in a queue to the right of the playfield. At any time, the player can swap the falling tetromino with the one in the hold box above the playfield, but a tetromino that has been swapped out cannot be immediately swapped back in. The gray blocks below the falling tetromino are the shadow, which shows where the tetromino will land.

A "line", or a complete row of 10 blocks across the well, will disappear, and everything above it moves down by one row. But if the well fills so high that a tetromino is placed entirely out of bounds or in a position such that the next tetromino does not have room to enter, the player "tops out" and the game is over.

## <a name="installing">Installing</a>

The executable included with the official distribution is designed for Microsoft Windows systems. Unzip it into a folder on your hard disk or USB memory card.

The files `lj.exe`, `lj.dat`, `ljblocks.bmp`, and `alleg42.dll` are required to play. (If you get an error about `alleg42.dll`, see the instructions on the LOCKJAW download page.) To start the game, run `lj.exe` in Windows Explorer.

The program should work on any PC running Microsoft Windows 98 or newer operating system with an 800x600 pixel display at 16-bit or greater color depth and DirectX 7 software installed. Its source code is portable to any platform that supports the [Allegro library](http://alleg.sourceforge.net/), but it has been tested only on Microsoft Windows because the author has access only to machines that run Microsoft Windows. If you manage to get it working on any other operating system, get in touch with the author.

To recompile the program, such as if you are testing a patch or porting it to another system:

1.  Install a GCC toolchain, such as MinGW and MSYS from [MinGW.org](http://www.mingw.org/MinGWiki/index.php/GettingStarted)
2.  Install Allegro 4.2.0 from [SourceForge.net](http://alleg.sourceforge.net/)
3.  Install JPGalleg 2.5 from [Enhanced Creations++](http://www.ecplusplus.com/index.php?page=projects&pid=1)
4.  Install DUMB 0.9.3 from [SourceForge.net](http://dumb.sourceforge.net/)  
    (Editor's note: DUMB has a dumb license, but section 6 allows relicensing under GNU GPL. Update: It appears [Debian got the author to relicense it zlib style](http://packages.debian.org/changelogs/pool/main/libd/libdumb/libdumb_0.9.3-5/libaldmb1.copyright).)
5.  Open a command prompt and <kbd>cd</kbd> to the folder containing the file `makefile`
6.  Type `make`

## <a name="skin">Skinning</a>

You can customize the appearance of LOCKJAW for PC by using Notepad or any other text editor to create a skin description that lists the images that shall be used. Images can be in Windows bitmap (`.bmp`), PC-Paintbrush (`.pcx`), Truevision TGA (`.tga`), or JFIF/JPEG (`.jpg`) format. All color depths should be supported. By default, this skin description should be placed in a file called `skin.ini`, in the same folder as `lj.ini`. (You can create and edit .ini files in Notepad or any other plain text editor.) LOCKJAW recognizes the following commands in a skin description:

<dl>

<dt>`ljblocksSRS=_image_`</dt>

<dd>A 192x96 or 192x144 pixel image with 24x24 pixel block images, used in bounding-box rotation systems (SRS and TOD M4). Rows 1 and 2 are for blocks in the well, and rows 3 and 4 are for the falling piece. If rows 5 and 6 are present, the shadow will use those; otherwise, it will use rows 3 and 4.</dd>

<dt>`ljblocksSega=_image_`</dt>

<dd>Same as `ljblocksSRS`, but used in other rotation systems.</dd>

<dt>`ljconnSRS=_image_`</dt>

<dd>A 192x192 pixel image with O tetrominoes of each color, which the loader cuts up to form tetrominoes with the blocks drawn connected, used for blocks in the well and the falling piece (not the shadow or empty areas of the well) in bounding-box rotation systems. If this file is not present, `ljblocksSRS` will be used instead.</dd>

<dt>`ljconnSega=_image_`</dt>

<dd>Same as `ljconnSRS`, but used in other rotation systems. If this file is not present, `ljblocksSega` will be used instead.</dd>

<dt>`color=_#RRGGBB_`</dt>

<dd>A 3- or 6-digit hexadecimal color (e.g. `#ABC` or `#D0FFE3`) for text in the menus and during the game. If not present, the game will use black.</dd>

<dt>`bgcolor=_#RRGGBB_`</dt>

<dd>A color for the menus' background. If not present, the game will use white.</dd>

<dt>`hilitecolor=_#RRGGBB_`</dt>

<dd>A color for the background of highlighted text. If not present, the game will use pale yellow (#FFC).</dd>

<dt>`pfcolor=_#RRGGBB_`</dt>

<dd>A color for text in front of the playfield. If not present, the game will use white.</dd>

<dt>`pfbgcolor=_#RRGGBB_`</dt>

<dd>An "average" color for the playfield, to be displayed behind the pause screen and during the game over animation. If not present, the game will use black.</dd>

<dt>`ljbg=_#RRGGBB_`</dt>

<dd>An 800x600 pixel image to be displayed behind the game. If this file is not 800x600 pixels, the image will be resized (sloppily) after being loaded. If this file is not present, the game will use a plain backdrop of the same color as `bgcolor`.</dd>

<dt>`bgm=_music file_`</dt>

<dd>A music file in Vorbis format (`.ogg`) or tracker format (`.mod`, `.s3m`, `.xm`, or `.it`) to be played during the game. You can create Vorbis format files by extracting audio from your CDs to `.wav` format using [CDex software](http://sourceforge.net/projects/cdexos/) and converting them with [OggDropXPd software](http://www.rarewares.org/ogg.html). You can download tracker format files from [The Mod Archive](http://www.modarchive.com/) or create them yourself using the [OpenMPT music editor](http://lpchip.com/modplug/viewtopic.php?t=18), the continuation of MODPlug Tracker. If this file is not present, the game will not play music.</dd>

</dl>

You can create more than one skin description file and then switch among them by using the following command in `lj.ini`:

<dl>

<dt>`Skin=_skin description file_`</dt>

<dd>A skin description file. If the skin is not present, the game uses the default file names (`ljblocks.bmp`, `ljblocks-sega.bmp`, `ljconn.bmp`, `ljconn-sega.bmp`, `ljbg.jpg`, and `bgm.s3m`).</dd>

</dl>

Example image and sound files are located in `lj-contrib.zip`, available from the [Skins section of the download page](http://www.pineight.com/lj/dl#contrib). You can also customize the sound effects by using [Allegro Grabber software](http://www.allegro.cc/depot/Grabber/) to edit `lj.dat`.

## <a name="controls">Controls</a>

### PC

The default controls in LOCKJAW for PC are as follows:

<dl>

<dt>← Shift left</dt>

<dd>Left arrow key</dd>

<dt>→ Shift right</dt>

<dd>Right arrow key</dd>

<dt>↓ Soft drop</dt>

<dd>Down arrow key</dd>

<dt><u>↓</u> Hard drop</dt>

<dd>Up arrow key</dd>

<dt>↰ Rotate left</dt>

<dd>Z, C</dd>

<dt>↱ Rotate right</dt>

<dd>X</dd>

<dt>↶ Rotate twice</dt>

<dd>W</dd>

<dt>↖ Hold piece</dt>

<dd>S</dd>

<dt>⇤ Shift far left</dt>

<dd>Q</dd>

<dt>⇥ Shift far right</dt>

<dd>E</dd>

<dt><u>↓</u> Firm drop</dt>

<dd>Enter</dd>

<dt>↖ Alt. hold piece</dt>

<dd>D</dd>

</dl>

Controls are configurable to the keyboard or any compatible joystick. (The key labeled "Item" is not yet used in the game.) From the main menu, choose "Game Keys..." and then press the keys in order as prompted. If you don't want to bind a key or button to a given function, press a key that you won't use. The key bindings are saved to the file `lj-keys.005`; if they become unusable, you can delete this file to reset them to the defaults.

Some controls during game play are hard-coded to keyboard keys:

*   Esc pauses and resumes the game. Holding Esc for one second stops the game and goes to a result screen.
*   [ (left bracket) starts and stops demo recording, and ] (right bracket) starts and stops demo playback. An icon for stop, record, or play appears in the upper left corner of the window. The demo is saved to a file called `demo.ljm`. **Caution:** Demos recorded on one version of LOCKJAW may not play correctly on another version. Recording another demo will overwrite the last demo, so make sure to rename the demo when you record a good one.

In the menus, Esc acts as Rotate Left, and Enter acts as Rotate Right. In all screens, Print Screen (F13) saves a copy of the display to the file `ljsnap.bmp`.

### GBA and DS

Controls in the Game Boy Advance and Nintendo DS versions are hardcoded as follows:

<dl>

<dt>← Shift left</dt>

<dd>Left on Control Pad</dd>

<dt>→ Shift right</dt>

<dd>Right on Control Pad</dd>

<dt>↓ Soft drop</dt>

<dd>Down on Control Pad</dd>

<dt><u>↓</u> Hard drop</dt>

<dd>Up on Control Pad</dd>

<dt>↰ Rotate left</dt>

<dd>B Button</dd>

<dt>↱ Rotate right</dt>

<dd>A Button</dd>

<dt>↖ Hold piece</dt>

<dd>L Button</dd>

</dl>

In the options menu, the B and A Buttons change screens, and the Start Button starts the game. Pressing A on the last screen of options also starts the game. Start pauses and resumes the game.

### Movement features

Unlike some other <abbr title="Soviet Mind Game">S.M.G.</abbr> implementations, LOCKJAW features Initial Actions. Holding a rotate button while a tetromino enters the playfield will cause the action to be performed the moment the tetromino enters. This is important for fast play. In addition, the hold key works at any time; if a piece is not falling yet, it will swap the hold piece with the next piece.

LOCKJAW also implements the so-called [Super Rotation System (SRS)](http://www.tetrisconcept.com/tetriswiki/SRS), which allows a tetromino to rotate around obstacles for more mobility across the pile. Some players abuse SRS by [rotating a piece in place repeatedly](http://infinitespin.ytmnd.com/), but this will result in poorer scores in timed gimmicks. To replace this behavior with the "Sega style" behavior used in Arika's _Tetris The Grand Master 3: Terror-Instinct_, change the [Options described below](#options).

## <a name="scoring">Scoring</a>

LOCKJAW allows the player to choose among several scoring methods. The terms "single", "double", "triple", and "home run" refer to clearing 1, 2, 3, or 4 lines with one tetromino. (A "home run" is called a "tetris" in some other games.) "T-spin" means rotating a T piece as it lands to fit into a tight space.

In LJ and TDS, making a bonus line clear when your last line clear was also bonus ("back-to-back homer" or "back-to-back T-spin") will produce extra points. Making a T-spin that does not clear a line has no effect on bonus state.

### LJ

As you clear lines, you also produce garbage that depends on the number of lines that you clear with a single tetromino. In single player mode, you earn 100 points per line cleared and 200 points per line of garbage, and in multiplayer mode (which is not yet implemented), the garbage will push up the blocks in another player's well.

<table summary="first row: column headings"><caption>LJ scoring method</caption>

<thead>

<tr>

<th></th>

<th>Lines</th>

<th>Garbage</th>

<th>Score</th>

<th>Bonus</th>

</tr>

</thead>

<tbody>

<tr>

<td>1 line</td>

<td>1</td>

<td>0</td>

<td>100</td>

<td>No</td>

</tr>

<tr>

<td>2 lines</td>

<td>2</td>

<td>1</td>

<td>400</td>

<td>No</td>

</tr>

<tr>

<td>3 lines</td>

<td>3</td>

<td>2</td>

<td>700</td>

<td>No</td>

</tr>

<tr>

<td>4 lines</td>

<td>4</td>

<td>4</td>

<td>1200</td>

<td>Yes</td>

</tr>

<tr>

<td>1 line T-spin</td>

<td>1</td>

<td>2</td>

<td>500</td>

<td>Yes</td>

</tr>

<tr>

<td>2 lines T-spin</td>

<td>2</td>

<td>4</td>

<td>1000</td>

<td>Yes</td>

</tr>

<tr>

<td>3 lines T-spin</td>

<td>3</td>

<td>6</td>

<td>1500</td>

<td>Yes</td>

</tr>

<tr>

<td>Back-to-back bonus</td>

<td></td>

<td>1</td>

<td>200</td>

<td>Yes</td>

</tr>

</tbody>

</table>

### TNT64

Based on the Fibonacci sequence.

<table summary="first row: column headings"><caption>TNT scoring method</caption>

<thead>

<tr>

<th>Lines</th>

<th>Score</th>

</tr>

</thead>

<tbody>

<tr>

<td>1</td>

<td>100</td>

</tr>

<tr>

<td>2</td>

<td>200</td>

</tr>

<tr>

<td>3</td>

<td>300</td>

</tr>

<tr>

<td>4</td>

<td>500</td>

</tr>

<tr>

<td>5</td>

<td>800</td>

</tr>

<tr>

<td>6</td>

<td>1300</td>

</tr>

<tr>

<td>7</td>

<td>2100</td>

</tr>

</tbody>

</table>

### Hotline

Only lines cleared on specific rows are worth points. There is no back-to-back bonus nor T-spin bonus.

<table summary="first row: column headings"><caption>Hotline scoring method</caption>

<tbody>

<tr>

<th>Row</th>

<th>Score</th>

</tr>

</tbody>

<tbody>

<tr>

<td>5</td>

<td>100</td>

</tr>

<tr>

<td>10</td>

<td>200</td>

</tr>

<tr>

<td>14</td>

<td>300</td>

</tr>

<tr>

<td>17</td>

<td>400</td>

</tr>

<tr>

<td>19</td>

<td>500</td>

</tr>

<tr>

<td>20</td>

<td>600</td>

</tr>

</tbody>

</table>

### TDS

Each line clear score is multiplied by a section number, computed by dividing the total number of lines cleared before this line by 10 and adding 1\. Thus, lines cleared later in the game when gravity is faster are worth more than lines cleared earlier. In TDS, the section number stops increasing after 190 lines.

<table summary="first row: column headings"><caption>TDS scoring method</caption>

<thead>

<tr>

<th></th>

<th>Lines</th>

<th>Score</th>

<th>Bonus</th>

</tr>

</thead>

<tbody>

<tr>

<td>1 line</td>

<td>1</td>

<td>100 * section</td>

<td>No</td>

</tr>

<tr>

<td>2 lines</td>

<td>2</td>

<td>300 * section</td>

<td>No</td>

</tr>

<tr>

<td>3 lines</td>

<td>3</td>

<td>500 * section</td>

<td>No</td>

</tr>

<tr>

<td>4 lines</td>

<td>4</td>

<td>800 * section</td>

<td>Yes</td>

</tr>

<tr>

<td>0 lines T-spin (wall kick)</td>

<td>0</td>

<td>100 * section</td>

<td></td>

</tr>

<tr>

<td>0 lines T-spin (no wall kick)</td>

<td>0</td>

<td>400 * section</td>

<td></td>

</tr>

<tr>

<td>1 line T-spin (wall kick)</td>

<td>1</td>

<td>200 * section</td>

<td>Yes</td>

</tr>

<tr>

<td>1 line T-spin (no wall kick)</td>

<td>1</td>

<td>800 * section</td>

<td>Yes</td>

</tr>

<tr>

<td>2 lines T-spin</td>

<td>2</td>

<td>1200 * section</td>

<td>Yes</td>

</tr>

<tr>

<td>3 lines T-spin</td>

<td>3</td>

<td>1600 * section</td>

<td>Yes</td>

</tr>

<tr>

<td>Back-to-back bonus</td>

<td></td>

<td>50% more</td>

<td>Yes</td>

</tr>

</tbody>

</table>

### NES

Each line clear is multiplied by the section number, as in TDS. The section number does not stop increasing. There is no back-to-back bonus nor T-spin bonus.

<table summary="first row: column headings"><caption>NES scoring method</caption>

<thead>

<tr>

<th>Lines</th>

<th>Score</th>

</tr>

</thead>

<tbody>

<tr>

<td>1</td>

<td>40 * section</td>

</tr>

<tr>

<td>2</td>

<td>100 * section</td>

</tr>

<tr>

<td>3</td>

<td>300 * section</td>

</tr>

<tr>

<td>4</td>

<td>1200 * section</td>

</tr>

</tbody>

</table>

### Drop scoring

Some games award extra points every time the tetromino lands if the player used soft drop or hard drop. LOCKJAW can simulate these.

<dl>

<dt>None</dt>

<dd>Award no points.</dd>

<dt>Continuous</dt>


CHANGES.txt
0.34a (2007-03-27)
  * PC: Fixed hidden level in Next at right mode (0.33 regression).
  * DS: Fixed hidden sprites due to bug in devkitARM R20 libnds
    (0.34 regression).
  * DS: New touch screen code should eliminate erroneous low notes
    played upon pen-up.

_____________________________________________________________________
0.34 (2007-03-23)
  * Fixed Move to Back piece set (0.14 regression: dang!).
  * Separate options for number of previewed pieces above shadow and
    outside field (requested by kiwibonga, caffeine, and Cubicz).
  * Option to disable IRS (requested by DIGITAL).
  * Garbage gimmicks (vs., drill 40, hr derby) replaced with new
    garbage option (requested by Cubicz).
  * In garbage style "Vs.", player can choose difficulty 1 through 4
    (requested by Caithness).
  * Option for line clear delay, independent of entry delay
    (requested by kiwibonga).
  * Garbage style Drill no longer automatically tops out on
    short playfields.
  * Garbage style Vs. generates garbage with SZSZ randomizer.
  * Frequency of Vs. w/Items bananas is independent of randomizer.
  * Fixed self-clearing garbage (0.33 regression).
  * GBA/DS: Fixed pause button hiding left and right walls
    (0.33 regression).
  * GBA/DS: R button also works as hold piece (requested by
    PetitPrince).

_____________________________________________________________________
0.33 (2007-03-08)
  * New speed curve Rhythm Zero is identical to Rhythm except it
    operates at 0G instead of 20G (requested by Cubicz and caffeine).
  * Speed curves that do not use level reset level to 0.
    Therefore, playing a game of Death then a game of Exponential
    on GBA/DS no longer shows the level that the player finished
    Death on as the level during Exponential.
  * Maximum well width increased to 12 (requested by Zed0).
  * Hold piece does not reset floor kicks, lock delay time in entry
    reset, or Rhythm's placement timer.
  * GBA/DS: More shared code moved to file "ljgbads.inc".
  * PC: Stops end-of-section sound when game is over
    (requested by Bloodstar).

_____________________________________________________________________
0.32a (2007-02-25)
  * PC: Fixed demo recording (0.32 regression; reported by Rosti LFC)
  * PC: Option to record all games from the start
    (requested by cdsboy, DIGITAL, Cubicz, and colour_thief).
  * PC: Added Play .ljm gimmick:
    Allegro file selector pops up, and the player can choose the
    filename of a demo to play back (requested by Lardarse).

_____________________________________________________________________
0.32 (2007-02-13)
  * Changed a CPU yield behavior that was affecting battery
    efficiency on the GBA and DS front ends (reported by Mighty Max).
  * Removed Low Rider gimmick in favor of option for well height
    (requested by Matthew).
  * Option for well width, like Shimizu's Tetris Semipro-68k.
    Combine this with a low well height to simulate BIG mode of
    TGM and Heboris (requested by Matthew).
  * Options for entry delay and sideways delay have "max" added to
    their description to clarify things (requested by kotetsu213 and
    Ezzelin).
  * All speed options given in Hz or G are given in both.
  * PC: Next above shadow works in all playfield positions (reported
    by caffeine).
  * GBA/DS: Added option for soft drop speed (requested by Ezzelin).
  * GBA/DS: Some shared code moved to file "ljgbads.inc".
  * GBA/DS: Unavailable hold piece is grayed out.
  * GBA/DS: Lock delay = no lock works (reported by Ezzelin).
  * GBA/DS: Fixed frames/ms display of delays (reported by Ezzelin).

_____________________________________________________________________
0.31 (2007-02-03)
  * Draws the score and next pieces before blitting the playfield
    in case a front-end draws the score or next pieces inside the
    playfield.
  * In Baboo!, speed curve Zero sets the level to the number of
    keypresses so far.
  * PC, GBA: Plays sound effect for end of section in speed curves
    that use sections (Master, Death, NES, Game Boy).
  * PC: Option to draw next pieces inside the playfield, above the
    shadow (requested by caffeine, cdsboy, and Cubicz).
  * GBA, DS: Blocks within a piece are drawn connected once they lock
    (TOD parity, requested by Lardarse and Bloodstar).
  * GBA, DS: Baboo! no longer double-counts console buttons passed
    through the joypad reading code.
  * README: "color=image" clone-and-hack error fixed (reported by
    Bloodstar).
  * README: Explains black rectangle in upper left corner as an
    icon for stop (requested by herc).
  * README: Explains controls on GBA and DS.

_____________________________________________________________________
0.30 (2007-02-01)
  * Fixed ARE display in debrief, which could cause a crash in some
    situations (0.26 regression?).
  * Garbage no longer self-clears in Cascade gravity (reported by
    Bloodstar).
  * In high gravity, land a T with one block over an overhang and
    rotate it once so that it falls and clears a line.  This is no
    longer counted as a T-spin.
  * During a replay, counts keypresses by the replay player.
  * PC: Scales undersized background images to the size of the window
    (requested by Bloodstar).
  * PC: If a piece enters during a skipped frame, a second upward
    trail is no longer drawn (reported by Bloodstar and caffeine).
  * PC: "Next at top" layout draws speedometer (requested by
    matt_hatter83).
  * PC: Trails option is saved properly (reported by caffeine).
  * PC: User can customize the game's text and background colors
    using skin.ini (requested by Bloodstar).
  * Added rudimentary port to Nintendo DS for people with MAX Media
    Dock, M3 Pro, SuperCard Rumble, or SLOT-1 cards, which can run
    DS homebrew but not GBA homebrew.
  * Moscow Nights removed from lj-contrib due to research into the

src/pcjoy.c
/* PC joystick code

Copyright (C) 2006 Damian Yerrick <tepples+lj@spamcop.net>

This work is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

Original game concept and design by Alexey Pajitnov.
The Software is not sponsored or endorsed by Alexey Pajitnov, Elorg,
or The Tetris Company LLC.

*/

#include "pcjoy.h"
#include <allegro.h>
#include <string.h>

extern const FONT *aver32, *aver16;
extern int bgColor, fgColor, hiliteColor;

static volatile int lastScancodePressed = -1;
static void (*oldKeyListener)(int scancode);
volatile int wantsClose = 0;
void ezPlaySample(const char *filename, int vol);

static void ljpcKeyListener(int scancode) {
  if (!(scancode & 0x80)) {
    lastScancodePressed = scancode;
  }
  if (oldKeyListener) {
    oldKeyListener(scancode);
  }
} END_OF_FUNCTION(ljpcKeyListener);


/**
 * Presses the escape key when the user clicks the close box.
 */
static void setWantsClose() {
  wantsClose = 1;
}

static int withJoystick = 0;

#define N_VKEYS 14

static const struct pkeyMapping defaultKeymappings[N_VKEYS] = {
  {-1, -1, KEY_UP},
  {-1, -1, KEY_DOWN},
  {-1, -1, KEY_LEFT},
  {-1, -1, KEY_RIGHT},
  {-1, -1, KEY_Z},
  {-1, -1, KEY_X},
  {-1, -1, KEY_S},
  {-1, -1, KEY_SPACE},
  {-1, -1, KEY_C},
  {-1, -1, KEY_W},
  {-1, -1, KEY_Q},
  {-1, -1, KEY_E},
  {-1, -1, KEY_ENTER},
  {-1, -1, KEY_D}
};

static struct pkeyMapping keymappings[N_VKEYS];

void getPkeyName(char *dst, int j, int s, int a) {
  if (j < 0) {
    usprintf(dst, "Key %d (%s)", a, scancode_to_name(a));
  } else if (s < 0) {
    usprintf(dst,
             "Joy %d button %s",
             j,
             joy[j].button[a].name);
  } else if (a < 0) {
    usprintf(dst,
             "Joy %d stick %s axis %s -",
             j,
             joy[j].stick[s].name,
             joy[j].stick[s].axis[~a].name);
  } else {
    usprintf(dst,
             "Joy %d stick %s axis %s +",
             j,
             joy[j].stick[s].name,
             joy[j].stick[s].axis[a].name);
  }
}

static int getPkeyState(int j, int s, int a) {
  int k;
  if (j < 0) {
    k = key[a];
  } else if (s < 0) {
    k = joy[j].button[a].b;
  } else if (a < 0) {
    k = joy[j].stick[s].axis[~a].d1;
  } else {
    k = joy[j].stick[s].axis[a].d2;
  }
  return k;
}

const char *const vkeyNames[] = {
  "Up",
  "Down",
  "Left",
  "Right",
  "Rotate Left",
  "Rotate Right",
  "Hold",
  "Item",
  "Alt. Rotate Left",
  "Rotate Left Twice",
  "Far Left",
  "Far Right",
  "Alt. Firm Drop",
  "Alt. Hold",
  "Macro G",
  "Macro H"
};

static int getVkeyState(int vkey) {
  int j = keymappings[vkey].joy;
  int s = keymappings[vkey].stick;
  int a = keymappings[vkey].axis;

  return getPkeyState(j, s, a);
}

LJBits readPad(void) {
  int keys = 0;
  poll_joystick();

  for (int i = 0;
       i < N_VKEYS;
       ++i) {
    if (getVkeyState(i)) {
      keys |= 1 << i;
    }
  }
  return keys;
}

LJBits menuReadPad(void) {
  if (key[KEY_ENTER]) {
    return VKEY_ROTR;
  } else if (key[KEY_ESC]) {
    return VKEY_ROTL;
  } else if (key[KEY_UP]) {
    return VKEY_UP;
  } else if (key[KEY_DOWN]) {
    return VKEY_DOWN;
  } else if (key[KEY_LEFT]) {
    return VKEY_LEFT;
  } else if (key[KEY_RIGHT]) {
    return VKEY_RIGHT;
  } else {
    return readPad();
  }
}

// These contain the states of ALL buttons on ALL
// joysticks 
static LJBits lastConfigButtons[8];
static LJBits lastConfigStickAxis[8];

static int newButton(int *outJ, int *outS, int *outA) {
  poll_joystick();
  int found = 0;

  if (lastScancodePressed >= 0) {
    *outJ = -1;
    *outS = -1;
    *outA = lastScancodePressed;
    lastScancodePressed = -1;
    return 1;
  }

  for (int j = 0;
       j < num_joysticks;
       ++j) {
    LJBits cur = 0;

    for (int b = 0; b < joy[j].num_buttons; ++b) {
      if (joy[j].button[b].b) {
        if (!(lastConfigButtons[j] & (1 << b)) && !found) {
          *outJ = j;
          *outS = -1;
          *outA = b;
          found = 1;
        }
        cur |= 1 << b;
      }
    }
    lastConfigButtons[j] = cur;
  }
  if (found) {
    return 1;
  }

  for (int j = 0;
       j < num_joysticks;
       ++j) {
    LJBits cur = 0;
    LJBits mask = 1;

    for (int s = 0; s < joy[j].num_sticks; ++s) {
      for (int a = 0; a < joy[j].stick[s].num_axis; ++a) {
        if (joy[j].stick[s].axis[a].d1) {
          if (!(lastConfigStickAxis[j] & mask) && !found) {
            *outJ = j;
            *outS = s;
            *outA = ~a;
            found = 1;
          }
          cur |= mask;
        }
        mask <<= 1;
        if (joy[j].stick[s].axis[a].d2) {
          if (!(lastConfigStickAxis[j] & mask) && !found) {
            *outJ = j;
            *outS = s;
            *outA = a;
            found = 1;
          }
          cur |= mask;
        }
        mask <<= 1;
      }
    }
    lastConfigStickAxis[j] = cur;
  }
  return found;
}

static void clearNewButton(void) {
  int j, s, a;
  while (newButton(&j, &s, &a));
}

void loadKeys(const char *filename) {
  PACKFILE *fp = pack_fopen(filename, "r");

  memcpy(keymappings, defaultKeymappings, sizeof(keymappings));
  if (fp) {
    for (unsigned int vkey = 0;
         vkey < N_VKEYS && !pack_feof(fp);
         ++vkey) {
      int j = pack_getc(fp);
      int s = pack_getc(fp);
      int a = pack_getc(fp);

      keymappings[vkey].joy = j;
      keymappings[vkey].stick = s;
      keymappings[vkey].axis = a;
    }
    pack_fclose(fp);
  }
}

void saveKeys(const char *filename) {
  PACKFILE *fp = pack_fopen(filename, "w");

  if (fp) {
    for (unsigned int vkey = 0;
         vkey < N_VKEYS && !pack_feof(fp);
         ++vkey) {
      pack_putc(keymappings[vkey].joy, fp);
      pack_putc(keymappings[vkey].stick, fp);
      pack_putc(keymappings[vkey].axis, fp);
    }
    pack_fclose(fp);
  }
}

#define VKEY_ROWHT 24
#define VKEY_TOP 120

void drawVkeyRow(int vkey, int hilite) {
  char name[256];
  int y = VKEY_TOP + vkey * VKEY_ROWHT;
  
  rectfill(screen,
           16, y, 479, y + VKEY_ROWHT - 1,
           hilite ? hiliteColor : bgColor);
  if (hilite) {
    rect(screen,
         16, y, 479, y + VKEY_ROWHT - 1,
         0);
  }
  getPkeyName(name,
              keymappings[vkey].joy,
              keymappings[vkey].stick,
              keymappings[vkey].axis);
  textout_ex(screen, aver16, vkeyNames[vkey], 24, y + 4, fgColor, -1);
  textout_ex(screen, aver16, name, 240, y + 4, fgColor, -1);
}

void configureKeys(void) {
  clear_to_color(screen, bgColor);
  textout_ex(screen, aver32, "LOCKJAW > Game Keys", 16, 32, fgColor, -1);
  textout_ex(screen, aver16,
             "Press the key that you want to use for the highlighted action, or Esc to return.",
             40, 80, fgColor, -1);

  // Draw each vkey's name
  for (int vkey = 0; vkey < N_VKEYS; ++vkey) {
    drawVkeyRow(vkey, vkey == 0);
  };

  clearNewButton();
  for (int vkey = 0; vkey < N_VKEYS && !wantsClose; ) {
    int j, s, a;

    if (keypressed()) {
      int scancode;
      ureadkey(&scancode);
      if (scancode == KEY_ESC) {
        vkey = N_VKEYS;
      }
    }
    if (vkey < N_VKEYS && newButton(&j, &s, &a)) {
      if (j >= 0 || s > 0 || a != KEY_ESC) {
        keymappings[vkey].joy = j;
        keymappings[vkey].stick = s;
        keymappings[vkey].axis = a;
        ezPlaySample("nextS_wav", 128);
      }

      // Redraw old vkey
      drawVkeyRow(vkey, 0);
      rest(250);

      // Draw new vkey
      ++vkey;
      if (vkey < N_VKEYS) {
        drawVkeyRow(vkey, 1);
      }
      
      // Ignore all other buttons pressed at the same time
      clearNewButton();
    }
    rest(30);
  }
  saveKeys("lj-keys.005");
}


void initKeys(void) {
  LOCK_FUNCTION(ljpcKeyListener);
  LOCK_VARIABLE(lastScancodePressed);
  LOCK_VARIABLE(oldKeyListener);
  LOCK_FUNCTION(setWantsClose);
  LOCK_VARIABLE(wantsClose);

  oldKeyListener = keyboard_lowlevel_callback;
  keyboard_lowlevel_callback = ljpcKeyListener;
  withJoystick = !install_joystick(JOY_TYPE_AUTODETECT);
  loadKeys("lj-keys.005");
  set_close_button_callback(setWantsClose);
}


